packages/apollo-tracing/src/index.ts

```diff
@@ -33,6 +33,97 @@ interface ResolverCall {
+  requestDidStart() {
+    let startWallTime: Date | undefined;
+    let endWallTime: Date | undefined;
+    let startHrTime: HighResolutionTime | undefined;
+    let duration: HighResolutionTime | undefined;
+    const resolverCalls: ResolverCall[] = [];
+
+    startWallTime = new Date();
+    startHrTime = process.hrtime();
+
+    return {
+      executionDidStart() {
+        // It's a little odd that we record the end time after execution rather
+        // than at the end of the whole request, but because we need to include
+        // our formatted trace in the request itself, we have to record it
+        // before the request is over! It's also odd that we don't do traces for
+        // parse or validation errors, but runQuery doesn't currently support
+        // that, as format() is only invoked after execution.
+        return () => {
+          duration = process.hrtime(startHrTime);
+          endWallTime = new Date();
+        };
+      },
+      willResolveField(...args) {
+        const [, , , info] = args;
+
+        const resolverCall: ResolverCall = {
+          path: info.path,
+          fieldName: info.fieldName,
+          parentType: info.parentType,
+          returnType: info.returnType,
+          startOffset: process.hrtime(startHrTime),
+        };
+
+        resolverCalls.push(resolverCall);
+
+        return () => {
+          resolverCall.endOffset = process.hrtime(startHrTime);
+        };
+      },
+      willSendResponse({ response }) {
+        // In the event that we are called prior to the initialization of
+        // critical date metrics, we'll return undefined to signal that the
+        // extension did not format properly. Any undefined extension
+        // results are simply purged by the graphql-extensions module.
+        if (
+          typeof startWallTime === 'undefined' ||
+          typeof endWallTime === 'undefined' ||
+          typeof duration === 'undefined'
+        ) {
+          return;
+        }
+
+        const extensions =
+          response.extensions || (response.extensions = Object.create(null));
+
+        if (typeof extensions.tracing !== 'undefined') {
+          throw new Error("The tracing information already existed.");
+        }
+
+        // Set the extensions.
+        extensions.tracing = {
+          version: 1,
+          startTime: startWallTime.toISOString(),
+          endTime: endWallTime.toISOString(),
+          duration: durationHrTimeToNanos(duration),
+          execution: {
+            resolvers: resolverCalls.map(resolverCall => {
+              const startOffset = durationHrTimeToNanos(
+                resolverCall.startOffset,
+              );
+              const duration = resolverCall.endOffset
+                ? durationHrTimeToNanos(resolverCall.endOffset) - startOffset
+                : 0;
+              return {
+                path: [...responsePathAsArray(resolverCall.path)],
```

Comment:
Any reason to return a different array here than what the function returns?

A fix in case this wasn't the intent:
```suggestion
                path: responsePathAsArray(resolverCall.path),
```

Reply:
This maintains the behavior from the original implementation and I didn't second-guess it.  I believe the `responsePathAsArray` (also, `graphql-js`'s `pathToArray`) returns a linked list so I'm assuming this was meant to clone the array (as the spread notation would do in this case).

Reply:
This is a reasonable thing to scrutinize, and I wish that a comment indicated the reason for this when it was introduced.  This code was originally imported from a now-defunct `apollo-cache-control-js` package (which I happen to still have cloned locally, but doesn't provide any more insight).  In the previous implementation, it was not using this technique, but it appears that this spread was added to _get tests passing_ in https://github.com/apollographql/apollo-server/commit/3523e4ae02ce6ed2c8c211f141bc7f87ee42474e within #1294.

Anyhow, since the intent was largely to maintain identical behavior to what existed before.  I'd rather not jeopardize the success of landing this large refactor based on further changing implementation details and instead just make the change to the integration surface (plugin vs. extension).